interfaces/system_key.go

```diff
@@ -20,28 +20,51 @@
 package interfaces
 
 import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
 	"os"
 	"path/filepath"
-
-	"gopkg.in/yaml.v2"
+	"strings"
 
 	"github.com/snapcore/snapd/dirs"
 	"github.com/snapcore/snapd/logger"
 	"github.com/snapcore/snapd/osutil"
 	"github.com/snapcore/snapd/release"
 )
 
+// ErrSystemKeyIncomparableVersions indicates that the system-key
+// on disk and the system-key calculated from generateSystemKey
+// have different inputs and are therefor incomparable.
+//
+// This means:
+// - "snapd" needs to re-generate security profiles
+// - "snap run" cannot wait for those security profiles
+var (
+	ErrSystemKeyIncomparableVersions = errors.New("system-key versions not comparable")
+	ErrSystemKeyMissing              = errors.New("system-key missing on disk")
+)
+
 // systemKey describes the environment for which security profiles
 // have been generated. It is useful to compare if the current
 // running system is similar enough to the generated profiles or
 // if the profiles need to be re-generated to match the new system.
+//
+// Note that this key gets generated on *each* `snap run` - so it
+// *must* be cheap to calculate it (no hashes of big binaries etc).
 type systemKey struct {
-	BuildID          string   `yaml:"build-id"`
-	AppArmorFeatures []string `yaml:"apparmor-features"`
-	NFSHome          bool     `yaml:"nfs-home"`
-	OverlayRoot      string   `yaml:"overlay-root"`
-	Core             string   `yaml:"core,omitempty"`
-	SecCompActions   []string `yaml:"seccomp-features"`
+	// IMPORTANT: when adding new inputs bump this version
+	Version string `json:"version"`
+
+	BuildIDs struct {
+		HostSnapd     string `json:"host_snapd"`
+		CoreSnapSnapd string `json:"core_snap_snapd"`
+	} `json:"build_ids"`
```

Comment:
was it discussed to use both like this?  it will provoke some more regens than strictly necessary

Reply:
I will added the following comment:
       // The build-ids of both the core snap snapd and the host
	// snapd are used as inputs. This may lead to more regens
	// than strictly necessary. The alternative is to try to
	// figure out which one will be run, this means to look
	// for re-exec behaviour and for potential SNAP_REEXEC=0
	// in the snapd service file or /etc/environment. So
	// this approach is conservative by looking at both
	// at the expense of potential extra re-gens.

Reply:
to be clear it means  that switching a system that does reexec, to SNAP_REEXEC=0 will not regen profiles (unless because of Version?), I suppose we don't really support turning off reexec except in a very controlled way (the fact that is just a envvar from that POV is a bit misleading)